include/shark/Data/Impl/Downloader.hpp

```diff
@@ -0,0 +1,285 @@
+	bool writeAll(const char* buffer, std::size_t buffersize)
+	{
+		std::size_t pos = 0;
+		while (pos < buffersize)
+		{
+			std::size_t n = write(buffer + pos, buffersize - pos);
+			if (n == 0) return false;
+			pos += n;
+		}
+		return true;
+	}
+
+private:
+	int m_handle;                      ///< POSIX socket handle
+};
+
+} // namespace detail
+
+/// \brief Download a document with the HTTP protocol.
+///
+/// \param  url       download URL, for example "www.shark-ml.org/index.html"
+/// \param  port      TCP/IP port, defaults to 80
+///
+/// The function returns the HTTP request body. In case of success this
+/// is the requested document. In case of an error the function throws
+/// an exception. Note that the function does not perform standard
+/// actions of web browsers, e.g., execute javascript or follow http
+/// redirects. All HTTP response status codes other than 200 are
+/// reported as failure to download the document and trigger an
+/// exception.
+std::string download(std::string const& url, unsigned short port = 80)
+{
+	// split the URL into domain and resource
+	std::size_t start = 0;
+	if (url.size() >= 7 && url.substr(0, 7) == "http://") start = 7;
+	std::size_t slash = url.find('/', start);
+	if (slash == std::string::npos) throw std::runtime_error("[download] invalid URL (failed to split into domain and resource)");
+	std::string domain = url.substr(start, slash - start);
+	std::string resource = url.substr(slash);
+
+	// open a TCP/IP socket connection
+	detail::Socket socket(domain, port);
+	std::string request = "GET " + resource + " HTTP/1.1\r\nhost: " + domain + "\r\n\r\n";
+	socket.writeAll(request.c_str(), request.size());
+
+	// http reply data
+	std::map<std::string, std::string> headers;
+	std::string body;
+
+	// parse http reply line
+	std::string reply = socket.readLine();
+	if (reply.size() < 12) throw std::runtime_error("[download] http protocol violation");
+	if (reply.substr(0, 9) != "HTTP/1.0 " && reply.substr(0, 9) != "HTTP/1.1 ") throw std::runtime_error("[download] http protocol violation");
+	if (reply.substr(9, 3) != "200") throw std::runtime_error("[download] failed with HTTP status " + reply.substr(9));
+
+	// parse http headers
+	while (true)
+	{
+		std::string h = socket.readLine();
+		if (h.empty()) break;
+		std::size_t colon = h.find(":");
+		if (colon == std::string::npos) throw std::runtime_error("[download] http protocol violation");
+		std::string tag = h.substr(0, colon);
+		// convert plain ASCII to lower case
+		for (std::size_t i=0; i<tag.size(); i++) if (tag[i] >= 65 && tag[i] <= 90) tag[i] += 32;
+		std::string value = h.substr(colon + 1);
+		while (! value.empty() && value[0] == ' ') value.erase(0, 1);
+		while (! value.empty() && value[value.size() - 1] == ' ') value.erase(value.size() - 1);
+		// convert plain ASCII to lower case
+		for (std::size_t i=0; i<value.size(); i++) if (value[i] >= 65 && value[i] <= 90) value[i] += 32;
```

Comment:
> std::tolower

Maybe I am overly pessimistic here, but I generally avoid functions that
depend on the "current locale". What if someone compiles this code with,
say, Greek, Arab, or Chinese locale? Will tolower then still do its job?
The problem is that the HTTP standard does not care for a locale.